README.md
# Maximized Window Fix for Visual Studio Code

<!-- ![Maximize Window](icon.png) -->

**Important:** The issue [seems to be fixed](https://github.com/microsoft/vscode/issues/85592#event-3884295282) in the current version of VS Code (since V1.51, October 2020), so this extension should no longer be required.  If you are on that version or later, the extension will not auto-run the fix and it will suggest uninstalling.  If you find you do still need it, you can run the fix manually by clicking on the status bar icon (☐), and/or set the `vscode-maximize-fix.force` extension setting described below to prevent the warning.


## Features

This extension works around [VS Code issue 85592](https://github.com/microsoft/vscode/issues/85592), where the maximized window is larger than the screen on some computers with mixed DPI monitors.  

The fix runs automatically when the extension starts, and a small small status bar icon (☐) shows feedback.  It works by finding VS Code windows (by default, those with titles ending with *Visual Studio Code* or *Visual Studio Code - Insiders*) and removes the `WS_CAPTION` window style. 

You should not have to manually run the fix but, if you choose to, the extension adds a command *Maximize Window Fix* (`vscode-maximize-fix.fix`) with default hot-key `Ctrl`+`Alt`+`Shift`+`M`, and you can also run this command by clicking on the status bar icon (☐).


## Requirements

This extension is only needed in affected versions running on Windows with the (default) setting `window.titleBarStyle` == `custom`.


## Extension Settings

Settings:

* `vscode-maximize-fix.auto` -- Automatically run the Maximize Window Fix when the extension starts. (default: true)
* `vscode-maximize-fix.statusbar` -- Shows a status bar icon (☐) to show feedback or to manually run the fix. (default: true)
* `vscode-maximize-fix.suffix-vscode` -- Window title suffix to match VS Code windows. (default: *Visual Studio Code*)
* `vscode-maximize-fix.suffix-vscode-insiders` -- Window title suffix to match VS Code Insiders windows. (default: *Visual Studio Code - Insiders*)
* `vscode-maximize-fix.force` -- Ignore checks for the right platform/configuration/version and always run the fix. Be warned that your title bar might go missing if you've changed the (default) setting `window.titleBarStyle` == `custom`. (default: false)


## Release Notes

See: [Change Log](CHANGELOG.md)


## Installation

To install the extension, click *Install* on:

* [Visual Studio Marketplace: Maximize Window Fix](https://marketplace.visualstudio.com/items?itemName=danielgjackson.vscode-maximize-fix)

Alternative installation methods:

* In *Visual Studio Code* open the *Extensions* panel (`Ctrl`+`Shift`+`X`), search for `Maximized Window Fix` and click *Install*
* Or, in *Visual Studio Code* press `Ctrl`+`P` and type: `ext install danielgjackson.vscode-maximize-fix`.
* Or, manually *Install from VSIX...* the ready-built `.vsix` file from: [GitHub releases](https://github.com/danielgjackson/vscode-maximize-fix/releases).

The extension is [open source](https://github.com/danielgjackson/vscode-maximize-fix/blob/master/LICENSE.txt) and the source code is available at:

* [GitHub page: danielgjackson/vscode-maximize-fix](https://github.com/danielgjackson/vscode-maximize-fix)

You can download the source with:

* `git clone https://github.com/danielgjackson/vscode-maximize-fix`
* Or, download a snapshot: `PowerShell -Command "& {Invoke-WebRequest https://github.com/danielgjackson/vscode-maximize-fix/archive/master.zip -o _temp.zip ; Expand-Archive _temp.zip ; move _temp\vscode-maximize-fix-master vscode-maximize-fix ; rmdir _temp ; del _temp.zip}"`
<!--
* Or, download a snapshot using Windows Subsystem for Linux: `wsl curl -L https://github.com/danielgjackson/vscode-maximize-fix/archive/master.zip -o _temp.zip ; unzip _temp.zip -d _temp ; mv _temp/vscode-maximize-fix-master vscode-maximize-fix ; rmdir _temp ; rm _temp.zip`
-->

You can build your own extension package `.vsix` file (*Node.js* must be installed first):

* `npm install -g vsce && npm install && vsce package`

<!--
To publish your extension:

* `vsce publish`

...if the Personal Access Token has expired, first [create a new PAT](https://code.visualstudio.com/api/working-with-extensions/publishing-extension#get-a-personal-access-token) and login with: `vsce login $PUBLISHER`
-->

As an alternative to using the extension, you can manually run the underlying executable `vscode-maximize-fix.exe` (you may specify custom window title suffixes as arguments):

* If already installed with the extension, located at: `%USERPROFILE%\.vscode\extensions\danielgjackson.vscode-maximize-fix-*`.
* Or, download the ready-built binary from: [GitHub releases](https://github.com/danielgjackson/vscode-maximize-fix/releases).
* Or, build from source using Microsoft Build Tools: `build.cmd`.
* Or, build from source by cross-compiling from Windows Subsystem for Linux (with a Ubuntu/Debian default distro): `wsl sudo apt install build-essential gcc-mingw-w64 && wsl make`.


CHANGELOG.md
# Change Log

## 1.5.0

- The underlying issue seems to be fixed in the current stable version of VS Code.  On a fixed version, the extension informs the user that it has not auto-run and suggests uninstalling.  It can still be run manually, and the version check can be bypassed with the `vscode-maximize-fix.force` extension setting.

## 1.4.0

- Added a check for the (default) setting `window.titleBarStyle` == `custom`.
- Created an option to ignore system/configuration checks and run anyway: `vscode-maximize-fix.force`.
- Set the toolbar icon's tooltip as the last status message.
- Fixed a reporting bug if more than one window was adjusted.
- Removed junk files from the .vsix package.

## 1.3.0

- Bumped version number to match executable release.
- Automatically runs on startup by default.
- Added configuration settings (auto-start, status bar, custom window title suffixes).
- Status now reports how many windows received the fix.

## 0.0.3

- Improved README.

## 0.0.2

- Window title filter matches VS Code Insiders builds too.

## 0.0.1

- Initial release


vscode-maximize-fix.c
// Tweaks VS Code window style to fix an issue where the maximized window is larger than the screen.
// Dan Jackson, 2020.

#include <windows.h>
#include <tchar.h>
#include <stdio.h>

// State for finding a window
typedef struct
{
	const TCHAR **suffix;
	int numSuffixes;
	int numMatches;
	int numChanges;
} find_state_t;

static BOOL AdjustWindowStyles(HWND hWnd)
{
	LONG style = GetWindowLong(hWnd, GWL_STYLE);
	
	// Don't adjust windows that have already had their style altered
	if (!(style & WS_CAPTION))
	{
		return FALSE;
	}
	
	// Remove the WS_CAPTION style
	style &= ~WS_CAPTION;
	SetWindowLong(hWnd, GWL_STYLE, style);
	
	// So that new styles take effect?
	ShowWindow(hWnd, SW_HIDE);
	ShowWindow(hWnd, SW_SHOW);
	
	// If already maximized, restore and maximize again
	if (style & WS_MAXIMIZE)
	{
		ShowWindow(hWnd, SW_RESTORE);
		ShowWindow(hWnd, SW_MAXIMIZE);
	}
	
	return TRUE;
}

// Callback helper to match on a window title's suffix
static BOOL CALLBACK enumFuncFindSuffix(HWND hWnd, LPARAM lparam)
{
	find_state_t *findState = (find_state_t *)lparam;
	if (IsWindowVisible(hWnd))
	{
		// Window's title text
		int length = GetWindowTextLength(hWnd);
		TCHAR *windowTitle = (TCHAR *)malloc((length + 1) * sizeof(TCHAR));
		GetWindowText(hWnd, windowTitle, length + 1);
		
		BOOL matched = FALSE;
		for (int i = 0; i < findState->numSuffixes; i++)
		{
			// Find last matching substring
			TCHAR *last = NULL;
			for (TCHAR *match, *search = windowTitle; (match = _tcsstr(search, findState->suffix[i])) != NULL; search++)
			{
				last = match;
			}
			
			// Only matched if it is the title's suffix
			matched = (last != NULL && (last - windowTitle) + _tcslen(findState->suffix[i]) >= _tcslen(windowTitle));
			if (matched) break;
		}
		
		// Adjust matching windows
		if (matched)
		{
			findState->numMatches++;
			if (AdjustWindowStyles(hWnd))
			{
				findState->numChanges++;
				_tprintf(TEXT("FIXED #%d/%d: %s\n"), findState->numChanges, findState->numMatches, windowTitle);
			}
			else
			{
				_tprintf(TEXT("SKIPPED #%d/%d: %s\n"), findState->numMatches - findState->numChanges, findState->numMatches, windowTitle);
			}
		}
		
		free(windowTitle);
	}
	return TRUE;
}

int main(int argc, TCHAR *argv[])
{
	find_state_t findState = {0};
	
	// Default window suffixes
	const TCHAR *defaultSuffixes[] = {
		TEXT("Visual Studio Code"),
		TEXT("Visual Studio Code - Insiders"),
	};
	findState.numSuffixes = sizeof(defaultSuffixes) / sizeof(TCHAR *);
	findState.suffix = defaultSuffixes;
	
	// Overridden by command line options
	if (argc > 1)
	{
		findState.numSuffixes = argc - 1;
		findState.suffix = (const TCHAR **)(argv + 1);
	}
	
	EnumWindows(enumFuncFindSuffix, (LPARAM)&findState);
	if (findState.numMatches == 0)
	{
		_tprintf(TEXT("ERROR: No windows found matching suffix(es): "));
		for (int i = 0; i < findState.numSuffixes; i++)
		{
			_tprintf(TEXT("%s\"%s\""), i > 0 ? TEXT(" | ") : TEXT(""), findState.suffix[i]);
		}
		_tprintf(TEXT("\n"));
		return 1;
	}
	return 0;
}


extension.js
const vscode = require('vscode');
const os = require('os')
const fs = require('fs');
const child_process = require('child_process');

const title = 'Maximize Window Fix';

// Assign this to the three-part version number that fixes the issue (the user is prompted to uninstall the extension).
const fixedVersionString = '1.51.0';

const suffixList = [
	'Visual Studio Code',
	'Visual Studio Code - Insiders',
];

function spawnProcess(context, suffixList) {
	let status = { fixed: 0, total: 0, exit: null, error: null };
	return new Promise((resolve, reject) => {
		try {
			let outBuffer = "";
			const options = {
				stdio: [
					null,
					'pipe',
					'pipe',
				]
			}
			
			// Executable that tweaks VS Code window style to fix the issue
			const command = context.asAbsolutePath('vscode-maximize-fix.exe');
			if (!fs.existsSync(command)) {
				throw `External program not found: ${command}`
			}
			const process = child_process.spawn(command, suffixList, options);
	
			process.stdout.on('data', (data) => {
				outBuffer += data;
				let idx;
				while ((idx = outBuffer.indexOf('\n')) >= 0) {
					const line = outBuffer.substr(0, idx).trim();
					outBuffer = outBuffer.substr(idx + 1);
					// "FIXED #1/1: (title) - Visual Studio Code"
					// "SKIPPED #1/2: (title) - Visual Studio Code"
					const parts = line.replace('#', '').replace('/', ' ').replace(':', ' ').split(' ', 3);
					if (parts.length == 3) {
						if (parts[0] == 'FIXED') status.fixed = parseInt(parts[1]);
						status.total = parseInt(parts[2]);
					} else {
						throw `Unexpected program output: ${data}`
					}
				}
			});
	
			process.stderr.on('data', (data) => {
				throw `Unexpected program error output: ${data}`
			});
	
			process.on('close', (code) => {
				status.exit = code;
				if (code) {
					throw `Unexpected program exit code: ${code}`
				}
				resolve(status);
			});
		} catch (e) {
			status.error = e;
			reject(status);
		}
	});
}

async function runFix(context) {
	const config = vscode.workspace.getConfiguration('vscode-maximize-fix');
	if (os.platform() !== 'win32' && !config.force) {
		vscode.window.showWarningMessage(`${title}: This extension can only work on Windows.`);
	} else {
		try {
			const windowConfig = vscode.workspace.getConfiguration('window');
			if (windowConfig.titleBarStyle === 'custom' || config.force) {
				//vscode.window.showInformationMessage(`${title}: Fixing maximized windows`);
				statusMessage(`Fixing...`);
				const result = await spawnProcess(context, suffixList);
				let message;
				if (result.fixed == 0) {
					message = `No new windows to fix (${result.total} already fixed).`;
				} else if (result.fixed == 1) {
					message = `Maximized window fix applied.`;
				} else {
					message = `Maximized window fix applied to ${result.fixed} windows.`;
				}
				statusMessage(message);
			} else {
				vscode.window.showWarningMessage(`${title}: This extension would only run properly with the setting: "window.titleBarStyle" == "custom". This check can be overridden with: "vscode-maximize-fix.force".`);
			}
		} catch (e) {
			const msg = `Error while fixing maximized windows${config.force ? ' (was forced to try)' : ''}: ${e.error}.`;
			statusMessage(msg);
			vscode.window.showErrorMessage(`${title}: ${msg}`);
		}
	}
}

let statusBarItem = null;

/**
 * @param {vscode.ExtensionContext} context
 */
function activate(context) {
	const commandId = 'vscode-maximize-fix.fix';
	
	console.log(`${title}: activate()`);

	const config = vscode.workspace.getConfiguration('vscode-maximize-fix');

	// Override built-in suffixes with config
	if (config['suffix-vscode'] != null) suffixList[0] = config['suffix-vscode'];
	if (config['suffix-vscode-insiders'] != null) suffixList[1] = config['suffix-vscode-insiders'];
	
	const disposable = vscode.commands.registerCommand(commandId, () => {
		runFix(context);
	});
	context.subscriptions.push(disposable);
	
	// Status bar item
	if (config.statusbar) {
		statusBarItem = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 105);
		statusBarItem.command = commandId;
		statusBarItem.tooltip = title;
		context.subscriptions.push(statusBarItem);
	}
	updateStatusBarItem();

	// If the issue is fixed, this will inform the user if they are using a fixed version
	let disableAutoRun = false;
	if (fixedVersionString && !config.force) {
		const fix = fixedVersionString.split('.').map(n => parseInt(n));
		const cur = vscode.version.split('.').map(n => parseInt(n));
		if (cur[0] > fix[0]
			|| (cur[0] == fix[0] && cur[1] > fix[1])
			|| (cur[0] == fix[0] && cur[1] == fix[1] && cur[2] >= fix[2])) {
				
			disableAutoRun = true;
			const infoLink = { 
				type: 'url', 
				//title: 'More Details',
				//url: 'https://github.com/danielgjackson/vscode-maximize-fix/blob/master/README.md',
				title: 'VS Code Issue 85592', 
				url: 'https://github.com/microsoft/vscode/issues/85592#event-3884295282',
			};
			const extensionSettings = {
				type: 'extension-settings',
				title: 'Extension Settings...',
			};
			vscode.window.showInformationMessage(
				`${title}: 
					Thanks for using this extension -- but you can probably uninstall it now! 
					The maximize issue should be fixed since V${fixedVersionString} (you are running V${vscode.version}). 
					${config.auto ? 'Because of this, the \'fix\' was not automatically applied. ' : ''}
					You can still apply it manually with the status bar icon (☐), or you can remove this warning by overriding the version check with the extension setting: "vscode-maximize-fix.force".
				`,
				infoLink,
				extensionSettings,
			).then((item) => {
				if (!item || !item.type) return;
				if (item.type == 'extension-settings') {
					vscode.commands.executeCommand('workbench.extensions.action.showExtensionsWithIds', ['danielgjackson.vscode-maximize-fix']);
				}
				if (item.type == 'url') {
					vscode.env.openExternal(vscode.Uri.parse(item.url));
				}
			});
		}
	}

	// Auto-run on start (if not disabled by running on a version with the bug fixed -- unless forced to do so!)
	if (config.auto && (!disableAutoRun || config.force)) {
		vscode.commands.executeCommand(commandId);
	}
}

let currentStatusMessage = null;
let currentStatusTimeout = null;

function updateStatusBarItem() {
	if (!statusBarItem) return;
	// https://code.visualstudio.com/api/references/icons-in-labels
	// $(chrome-maximize) $(window)
	let text = `$(chrome-maximize)`;
	if (currentStatusMessage) {
		text = text + ' ' + currentStatusMessage;
	}
	statusBarItem.text = text;
	// Leave tooltip as last message
	if (currentStatusMessage && statusBarItem.tooltip != currentStatusMessage) {
		statusBarItem.tooltip = currentStatusMessage;
	}
	statusBarItem.show();
}

function statusMessage(message) {
	const timeout = 2000;
	if (currentStatusTimeout) {
		clearTimeout(currentStatusTimeout);
		currentStatusTimeout = null;
	}
	if (!statusBarItem) {
		if (message) {
			vscode.window.setStatusBarMessage(message, timeout);
		}
		return;
	}
	currentStatusMessage = message;
	updateStatusBarItem();
	currentStatusTimeout = setTimeout(() => {
		statusMessage(null);
	}, timeout);
}

// this method is called when your extension is deactivated
function deactivate() {
	console.log(`${title}: deactivate()`);
	statusBarItem = null;
}

module.exports = {
	activate,
	deactivate
}
